[TOC]

# High-throughput bacterial cultivation and identification for the plant root microbiome (Culturome)

This is the detail high-throughput bacterial culturing protocols for our previous published *Nature Biotechnology* [1]. Now it is preparing for *Nature Protocols* named "High-throughput bacterial cultivation and identification for the plant root microbiome".

The overview of the workflow is in the following figures.

![image](http://210.75.224.110/github/Culturome/script/fig/fig1.png)

**Fig. 1 | Overview of high-throughput bacterial cultivation and identification system.**

![image](http://210.75.224.110/github/Culturome/script/fig/fig4.png)

**Fig. 4 | The work flow of informatic analysis to identify cultivated bacteria**. a, The procedure from raw data to clean sequences of the amplified region. The amplified clean sequences are obtained by merging paired-end reads, demultiplexing plate and well barcodes, and removing the sequences of PCR primers. b, The work flow to identify ASV and taxonomy. Clean sequences are dereplicated by VSEARCH, denoised by USEARCH to identify ASV sequences, then classified by RDP database for taxonomy. c, Sequence and taxonomy information of cultivated bacteria in 96-well microtiter plates. The ASV table shows bacterial 16S rRNA sequence information of each well in 96-well microtiter plate. The annotated ASV list shows the information of wells with bacteria belonging to corresponding ASV, including purity, sequence count, and taxonomy.

![image](http://210.75.224.110/github/Culturome/script/fig/fig5.png)

**Fig. 5 | Anticipated results for root bacterial cultivation (e.g. Oryza sativa, L).**
a. The bar plot showing the distribution of read counts generated by two-side barcode system in each plate. Reads count of amplified 16S rRNA gene sequences in each plate of cultivated bacteria reveal the depth and evenness of Illumina sequencing. n = 45. b. The read counts in negative (sterile water as PCR template) and positive controls (E. coli DNA as PCR template). c. A rarefaction curve of ASVs according to the number of wells containing bacteria. The curve reach plateau stage indicating that the number of plates in high-throughput bacterial isolation is sufficient. d. Distribution of wells containing different numbers of ASVs or genera. Because pure bacteria may contain 16S rRNA gene variations resulting in different ASVs, wells containing more one ASVs belonging to single genus may be pure. e. An example of purity of wells containing cultivated bacteria in 96-well microtiter plate. g. Cladogram showing the taxonomic distribution and occurrence frequency of cultivated bacteria. The inner ring represents the dereplicated ASVs from cultivated root bacteria. Heat map in the outer ring represent the log2 transformed number of cultivated bacterial isolates belong to the corresponding ASV.

## Equipment

Personal computer running with 64-bit Linux Ubuntu 16.04+ or CentOS 7.5+, at least dual-core 2.4 GHz, 4 GB of RAM (8 GB preferred) and 30 GB space. The analysis pipeline is mainly based on QIIME[2], USEARCH[3], VSEARCH[4] and home-made scripts (Culturome pipeline). All the home-made scripts and manual is deposit in https://github.com/YongxinLiu/Culturome. The dependency databases is RDP[5]. For convenient all users to use it, the alternative install methods including VirtualBox image (alternative Conda and Docker). Additionally, a graphic user interface Webserver will be construct for user to analysis culturomics data online.

### Computer requirements

- Operational system requirements: 64-bit Linux, such as Ubuntu 16.04+ or CentOS 7.5+.
- Hardware requirements: > 4 GB of RAM and 30 GB free disk space.

### Software

- Conda 4.8.3, software management system (https://repo.anaconda.com/miniconda/).
- R (https://cran.r-project.org). The analysis was tested using R 3.5.1. Packages include ggplot243, pheatmap44, dplyr, stringr should be installed. 
- QIIME version 1.9.1[2] http://qiime.org/ for split barcodes
- USEARCH v10.0.240[3] http://www.drive5.com/usearch
- VSEARCH v.2.7.1[4] or later https://github.com/torognes/vsearch
- GraPhlAn 1.1.3-1[6], for visualizing taxonomic tree
- Culturome 1.0 analysis pipeline, related scripts, and example results are available at https://github.com/YongxinLiu/Culturome

### Input data files
- Sequencing Data：The raw sequence data used in this paper have been deposited in the Genome Sequence Archive (GSA)[7] in BIG Data Center[8], under accession numbers CRA002517. 
- RDP train set 16[5] as taxonomy database. The USEARCH formatted taxonomy is in http://www.drive5.com/sintax.


## Equipment setup

For Linux system, conda (https://docs.conda.io/en/latest/miniconda.html) is basic for setup the pipeline. We tested the pipeline on Ubuntu 16.04 and 18.04.

- Conda 4.8.3+ (https://docs.conda.io/en/latest/miniconda.html). Download and install the latest version, and config the Bioconda50 channel for easy-install software in biology. 


    wget -c https://repo.continuum.io/miniconda/Miniconda2-latest-Linux-x86_64.sh
    bash Miniconda2-latest-Linux-x86_64.sh
    conda config --add channels bioconda


- Install QIIME


    conda install qiime=1.9.1
    
If can't install it due to can't solving the environment, try to create a new environment.
    
	conda create -n qiime1 python=2.7 qiime matplotlib=1.4.3 mock nose -c bioconda
	conda activate qiime1

More install methods or troubleshooting in http://qiime.org/install/install.html

If you are windows user, a QIIME Virtual Box can be downloaded from http://qiime.org/install/virtual_box.html, and following the instructions.

- Download 'Cultrome' pipeline and add to environment


    # Download
    wget -c https://github.com/YongxinLiu/Culturome/archive/master.zip
    unzip master.zip
    cd Culturome-master
    export PATH=`pwd`/script:$PATH
    chmod +x script/*
    
- USEARCH 10.0

Please visit http://www.drive5.com/usearch/download.html, and download [usearch10.0.240_i86linux32.gz](https://drive5.com/downloads/usearch10.0.240_i86linux32.gz), rename as "usearch", and put it in "script" directory.

Add executable permissions

    wget -c https://drive5.com/downloads/usearch10.0.240_i86linux32.gz -O script/usearch.gz
    gunzip script/usearch.gz
    chmod +x script/usearch

- VSEARCH 2.14.2

Visiting https://github.com/torognes/vsearch. Following the "Download and install"

    wget https://github.com/torognes/vsearch/archive/v2.14.2.tar.gz
    tar xzf v2.14.2.tar.gz
    cd vsearch-2.14.2
    ./autogen.sh
    ./configure
    make
    # as root or sudo make install
    sudo make install  

- [GraPhlAn 1.1.3-1](https://bitbucket.org/nsegata/graphlan)


    conda install graphlan

- Input data

Data were recommended deposited in GSA[7] of BIGD[8] (gsa.big.ac.cn), which greatly saving our downloading time. Any public data center, such as National Center for Biotechnology Information (NCBI)51/ European Bioinformatics Institute (EBI)52, can provide the download links in online analysis mode.
Here we using CRR127980 as example to test the pipeline and overview the results.


    # Download CRR127980 and rename as L1
    wget -c ftp://download.big.ac.cn/gsa3/CRA002517/CRR127980/CRR127980_f1.fq.gz -O L1_1.fq.gz
    wget -c ftp://download.big.ac.cn/gsa3/CRA002517/CRR127980/CRR127980_r2.fq.gz -O L1_2.fq.gz
    # decompress
    gunzip *.gz


- Taxonomy database

USEARCH formatted taxonomy is in http://www.drive5.com/sintax.

    wget -c http://www.drive5.com/sintax/rdp_16s_v16_sp.fa.gz
    gunzip *.gz

## Analysis protocols

If you only one library, please following the instruction from `pipeline.md`, and it is easier to understand.

If you have more than one library. Please following the instruction from `pipeline_multiple.md`, and it calls the loop method for batch processing.

## Reference

1. Jingying Zhang, Yong-Xin Liu, Na Zhang, Bin Hu, Tao Jin, Haoran Xu, Yuan Qin, Pengxu Yan, Xiaoning Zhang, Xiaoxuan Guo, Jing Hui, Shouyun Cao, Xin Wang, Chao Wang, Hui Wang, Baoyuan Qu, Guangyi Fan, Lixing Yuan, Ruben Garrido-Oter, Chengcai Chu & Yang Bai. NRT1.1B is associated with root microbiota composition and nitrogen use in field-grown rice. Nature Biotechnology. 2019, 37: 676-684. doi:10.1038/s41587-019-0104-4
2. J. Gregory Caporaso, Justin Kuczynski, Jesse Stombaugh, Kyle Bittinger, Frederic D. Bushman, Elizabeth K. Costello, Noah Fierer, Antonio Gonzalez Peña, Julia K. Goodrich, Jeffrey I. Gordon, Gavin A. Huttley, Scott T. Kelley, Dan Knights, Jeremy E. Koenig, Ruth E. Ley, Catherine A. Lozupone, Daniel McDonald, Brian D. Muegge, Meg Pirrung, Jens Reeder, Joel R. Sevinsky, Peter J. Turnbaugh, William A. Walters, Jeremy Widmann, Tanya Yatsunenko, Jesse Zaneveld & Rob Knight. QIIME allows analysis of high-throughput community sequencing data. Nature Methods. 2010, 7: 335-336. doi:10.1038/nmeth.f.303
3. Robert C. Edgar. Search and clustering orders of magnitude faster than BLAST. Bioinformatics. 2010, 26: 2460-2461. doi:10.1093/bioinformatics/btq461
4. Torbjørn Rognes, Tomáš Flouri, Ben Nichols, Christopher Quince & Frédéric Mahé. VSEARCH: a versatile open source tool for metagenomics. PeerJ. 2016, 4: e2584. doi:10.7717/peerj.2584
5. James R. Cole, Qiong Wang, Jordan A. Fish, Benli Chai, Donna M. McGarrell, Yanni Sun, C. Titus Brown, Andrea Porras-Alfaro, Cheryl R. Kuske & James M. Tiedje. Ribosomal Database Project: data and tools for high throughput rRNA analysis. Nucleic Acids Research. 2014, 42: D633-D642. doi:10.1093/nar/gkt1244
6. Francesco Asnicar, George Weingart, Timothy L. Tickle, Curtis Huttenhower & Nicola Segata. Compact graphical representation of phylogenetic data and metadata with GraPhlAn. PeerJ. 2015, 3: e1029. doi:10.7717/peerj.1029
7. Yanqing Wang, Fuhai Song, Junwei Zhu, Sisi Zhang, Yadong Yang, Tingting Chen, Bixia Tang, Lili Dong, Nan Ding, Qian Zhang, Zhouxian Bai, Xunong Dong, Huanxin Chen, Mingyuan Sun, Shuang Zhai, Yubin Sun, Lei Yu, Li Lan, Jingfa Xiao, Xiangdong Fang, Hongxing Lei, Zhang Zhang & Wenming Zhao. GSA: Genome Sequence Archive*. Genomics, Proteomics & Bioinformatics. 2017, 15: 14-18. doi:10.1016/j.gpb.2017.01.001
8. National Genomics Data Center Members Partners. Database Resources of the National Genomics Data Center in 2020. Nucleic Acids Research. 2020, 48: D24-D33. doi:10.1093/nar/gkz913
9. Eric W Sayers, Jeff Beck, J Rodney Brister, Evan E Bolton, Kathi Canese, Donald C Comeau, Kathryn Funk, Anne Ketter, Sunghwan Kim, Avi Kimchi, Paul A Kitts, Anatoliy Kuznetsov, Stacy Lathrop, Zhiyong Lu, Kelly McGarvey, Thomas L Madden, Terence D Murphy, Nuala O’Leary, Lon Phan, Valerie A Schneider, Françoise Thibaud-Nissen, Bart W Trawick, Kim D Pruitt & James Ostell. Database resources of the National Center for Biotechnology Information. Nucleic Acids Research. 2020, 48: D9-D16. doi:10.1093/nar/gkz899
10. Charles E Cook, Oana Stroe, Guy Cochrane, Ewan Birney & Rolf Apweiler. The European Bioinformatics Institute in 2020: building a global infrastructure of interconnected data resources for the life sciences. Nucleic Acids Research. 2019, 48: D17-D23. doi:10.1093/nar/gkz1033

